packages/react/runtime/src/renderToOpcodes/index.ts

```diff
@@ -241,11 +249,28 @@ function _renderToString(
     rendered = isTopLevelFragment ? rendered.props.children : rendered;
 
     // Recurse into children before invoking the after-diff hook
-    _renderToString(rendered, context, isSvgMode, selectValue, vnode, opcodes);
-    if (afterDiff) afterDiff(vnode);
-    vnode[PARENT] = undefined;
+    try {
+      _renderToString(rendered, context, isSvgMode, selectValue, vnode, opcodes, opcodes.length);
+    } catch (e) {
+      if (e && typeof e === 'object' && e.then && component && /* _childDidSuspend */ component.__c) {
+        component.setState({ /* _suspended */ __a: true });
+
+        if (component[DIRTY]) {
+          rendered = renderClassComponent(vnode, context);
+          component = vnode[COMPONENT];
+
+          opcodes.length = opcodesLength;
+          _renderToString(rendered, context, isSvgMode, selectValue, vnode, opcodes, opcodes.length);
+        }
+      } else {
+        throw e;
```

Comment:
If we `throw e` from here, I think we still need to use `mainThreadLazy` to catch this error to avoid uncaught exception in main thread.